### PR TITLE
lightning: fix incorrect _tidb_rowid allocator value after import for table with AUTO_ID_CACHE=1

### LINKED ISSUE #46100: AUTO_INCREMENT increase too large after importing NONCLUSTERED AUTO_ID_CACHE=1 table by lightning
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
Following is schema that will import by lighting
```
cat smalldb.t1-schema.sql
CREATE TABLE `t1` (
    `cacheid` int(11) NOT NULL AUTO_INCREMENT,
    `rk` varchar(16) NOT NULL,
    `cfq` varchar(10) NOT NULL,
    `ts` bigint(20) NOT NULL,
    `value` varchar(2048) DEFAULT NULL,
    PRIMARY KEY(`cacheid`) NONCLUSTERED)ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_ID_CACHE 1;
```
After import finished, hit following error when try to insert new rows.
```
mysql> show table t1 next_row_id;
+---------+------------+-------------+--------------------+----------------+
| DB_NAME | TABLE_NAME | COLUMN_NAME | NEXT_GLOBAL_ROW_ID | ID_TYPE        |
+---------+------------+-------------+--------------------+----------------+
| smalldb | t1         | _tidb_rowid |                  1 | _TIDB_ROWID    |
| smalldb | t1         | _tidb_rowid |          577898371 | AUTO_INCREMENT |
+---------+------------+-------------+--------------------+----------------+
2 rows in set (0.21 sec)

mysql> select count(*) from t1;
+----------+
| count(*) |
+----------+
|  3960000 |
+----------+
1 row in set (0.78 sec)

mysql> show create table t1;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                                                                              |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `cacheid` int(11) NOT NULL AUTO_INCREMENT,
  `rk` varchar(16) NOT NULL,
  `cfq` varchar(10) NOT NULL,
  `ts` bigint(20) NOT NULL,
  `value` varchar(2048) DEFAULT NULL,
  PRIMARY KEY (`cacheid`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=577898371 /*T![auto_id_cache] AUTO_ID_CACHE=1 */ |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.20 sec)

mysql> select max(cacheid) from t1;
+--------------+
| max(cacheid) |
+--------------+
|      3960000 |
+--------------+
1 row in set (0.20 sec)

mysql> insert into t1(rk,cfq,ts) values('rk_txxxxxxxxx','cfg_xxxxx',7);
ERROR 1062 (23000): Duplicate entry '?' for key 't1.PRIMARY'
```


### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
v7.3.0

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #46100

Problem Summary:

since TiDB v6.5.0, we have separate id allocator for `_tidb_rowid` and `auto_increment` id when `AUTO_ID_CACHE=1`, see https://github.com/pingcap/tidb/pull/39041, and `ALTER TABLE x AUTO_INCREMENT=x` only works on the allocator of `auto_increment` in this case, but if the table have implicit `_tidb_rowid`, we should update it's base too to make sure new insert works, else it will report `duplicate entry`

### What is changed and how it works?
- rebase on both allocators when needed after import

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/common/common.go [hunk 1]
```diff
@@ -89,20 +107,29 @@ func getGlobalAutoIDAlloc(store kv.Storage, dbID int64, tblInfo *model.TableInfo
...
+	// Since TiDB 6.5.0, row ID and auto ID are using different allocators when tblInfo.SepAutoInc is true
 	switch {
 	case hasRowID || hasAutoIncID:
-		return autoid.NewAllocator(store, dbID, tblInfo.ID, tblInfo.IsAutoIncColUnsigned(),
-			autoid.RowIDAllocType, noCache, tblVer), nil
+		allocators := make([]autoid.Allocator, 0, 2)
+		if tblInfo.SepAutoInc() && hasAutoIncID {
+			allocators = append(allocators, autoid.NewAllocator(store, dbID, tblInfo.ID, tblInfo.IsAutoIncColUnsigned(),
+				autoid.AutoIncrementType, noCache, tblVer))
+		}
```
Comment (D3Hunter) [inline br/pkg/lightning/common/common.go:125]: coverage false alarm, those lines has covered by TestAllocGlobalAutoID

**Review by GMHDBJD — APPROVED** [review]
lgtm

File: br/pkg/lightning/common/common.go [hunk 2]
```diff
@@ -52,24 +52,42 @@ var DefaultImportVariablesTiDB = map[string]string{
...
+			return err
+		}
+	}
+	return nil
 }
 
-func getGlobalAutoIDAlloc(store kv.Storage, dbID int64, tblInfo *model.TableInfo) (autoid.Allocator, error) {
+// GetGlobalAutoIDAlloc returns the autoID allocators for a table.
+// export it for testing.
+func GetGlobalAutoIDAlloc(store kv.Storage, dbID int64, tblInfo *model.TableInfo) ([]autoid.Allocator, error) {
```
Comment (lance6716) [inline br/pkg/lightning/common/common.go:90]: nit: we can implement a wrapper that expose one `Allocator` interface but it's two inner allocators.

**Review by lance6716 — COMMENTED** [review]
lgtm

File: br/pkg/lightning/common/common.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/common/common.go:90]: I'll comment the lgtm bot command after you respond that this suggestion will be included in this PR or not.

Comment (D3Hunter) [reply br/pkg/lightning/common/common.go:90]: will not do it in this pr

**Comment by D3Hunter** [comment]
@lance6716 @GMHDBJD move the location of rebase id, ptal https://github.com/pingcap/tidb/pull/46171/commits/f1ec6723e8bd408efcf8115122a82c141aa4e34c

**Comment by lance6716** [comment]
/hold

wait the BR test

**Comment by D3Hunter** [comment]
ci pass https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_br_integration_test/detail/pull_br_integration_test/332/pipeline

**Pull request merged**